File: examples/java/transaction/src/main/java/dev/angzarr/examples/transaction/TransactionService.java
package dev.angzarr.examples.transaction;

import com.google.protobuf.Any;
import com.google.protobuf.InvalidProtocolBufferException;
import examples.Domains.*;
import io.grpc.Status;
import io.grpc.stub.StreamObserver;
import dev.angzarr.BusinessLogicGrpc;
import dev.angzarr.BusinessResponse;
import dev.angzarr.CommandBook;
import dev.angzarr.ContextualCommand;
import dev.angzarr.EventBook;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * gRPC service implementation for transaction business logic.
 */
public class TransactionService extends BusinessLogicGrpc.BusinessLogicImplBase {
    private static final Logger logger = LoggerFactory.getLogger(TransactionService.class);

    private final TransactionLogic logic;

    public TransactionService(TransactionLogic logic) {
        this.logic = logic;
    }

    @Override
    public void handle(ContextualCommand request, StreamObserver<BusinessResponse> responseObserver) {
        try {
            EventBook events = processCommand(request);
            BusinessResponse response = BusinessResponse.newBuilder()
                .setEvents(events)
                .build();
            responseObserver.onNext(response);
            responseObserver.onCompleted();
        } catch (CommandValidationException e) {
            responseObserver.onError(Status.fromCode(e.getStatusCode())
                .withDescription(e.getMessage())
                .asRuntimeException());
        } catch (InvalidProtocolBufferException e) {
            responseObserver.onError(Status.INVALID_ARGUMENT
                .withDescription("Failed to parse command: " + e.getMessage())
                .asRuntimeException());
        } catch (Exception e) {
            logger.error("Unexpected error processing command", e);
            responseObserver.onError(Status.INTERNAL
                .withDescription("Internal error: " + e.getMessage())
                .asRuntimeException());
        }
    }

    private EventBook processCommand(ContextualCommand request)
            throws CommandValidationException, InvalidProtocolBufferException {
        CommandBook cmdBook = request.getCommand();
        EventBook priorEvents = request.getEvents();

        if (cmdBook == null || cmdBook.getPagesList().isEmpty()) {
            throw CommandValidationException.invalidArgument("CommandBook has no pages");
        }

        var cmdPage = cmdBook.getPages(0);
        if (!cmdPage.hasCommand()) {
            throw CommandValidationException.invalidArgument("Command page has no command");
        }

        TransactionState state = logic.rebuildState(priorEvents);
        Any command = cmdPage.getCommand();
        String typeUrl = command.getTypeUrl();

        EventBook result;
        if (typeUrl.endsWith("CreateTransaction")) {
            CreateTransaction cmd = command.unpack(CreateTransaction.class);
            result = logic.handleCreateTransaction(state, cmd.getCustomerId(), cmd.getItemsList());
        } else if (typeUrl.endsWith("ApplyDiscount")) {
            ApplyDiscount cmd = command.unpack(ApplyDiscount.class);
            result = logic.handleApplyDiscount(state, cmd.getDiscountType(), cmd.getValue(), cmd.getCouponCode());
        } else if (typeUrl.endsWith("CompleteTransaction")) {
            CompleteTransaction cmd = command.unpack(CompleteTransaction.class);
            result = logic.handleCompleteTransaction(state, cmd.getPaymentMethod());
        } else if (typeUrl.endsWith("CancelTransaction")) {
            CancelTransaction cmd = command.unpack(CancelTransaction.class);
            result = logic.handleCancelTransaction(state, cmd.getReason());
        } else {
            throw CommandValidationException.invalidArgument("Unknown command type: " + typeUrl);
        }

        return result.toBuilder()
            .setCover(cmdBook.getCover())
            .build();
    }
}

File: examples/kotlin/transaction/src/main/kotlin/dev/angzarr/examples/transaction/TransactionService.kt
package dev.angzarr.examples.transaction

import dev.angzarr.BusinessLogicGrpcKt
import dev.angzarr.BusinessResponse
import dev.angzarr.ContextualCommand
import examples.Domains.*
import io.grpc.Status

/**
 * gRPC service adapter for transaction business logic.
 * Thin wrapper that delegates to TransactionLogic and maps exceptions to gRPC status codes.
 */
class TransactionService(
    private val logic: TransactionLogic
) : BusinessLogicGrpcKt.BusinessLogicCoroutineImplBase() {

    override suspend fun handle(request: ContextualCommand): BusinessResponse {
        val cmdBook = request.command
        val priorEvents = request.events

        if (cmdBook.pagesList.isEmpty()) {
            throw Status.INVALID_ARGUMENT
                .withDescription("CommandBook has no pages")
                .asRuntimeException()
        }

        val cmdPage = cmdBook.pagesList[0]
        val cmd = cmdPage.command ?: throw Status.INVALID_ARGUMENT
            .withDescription("Command page has no command")
            .asRuntimeException()

        val state = logic.rebuildState(priorEvents)

        try {
            val eventBook = when {
                cmd.`is`(CreateTransaction::class.java) -> {
                    val c = cmd.unpack(CreateTransaction::class.java)
                    logic.handleCreateTransaction(state, c.customerId, c.itemsList)
                }
                cmd.`is`(ApplyDiscount::class.java) -> {
                    val c = cmd.unpack(ApplyDiscount::class.java)
                    logic.handleApplyDiscount(state, c.discountType, c.value, c.couponCode)
                }
                cmd.`is`(CompleteTransaction::class.java) -> {
                    val c = cmd.unpack(CompleteTransaction::class.java)
                    logic.handleCompleteTransaction(state, c.paymentMethod)
                }
                cmd.`is`(CancelTransaction::class.java) -> {
                    val c = cmd.unpack(CancelTransaction::class.java)
                    logic.handleCancelTransaction(state, c.reason)
                }
                else -> throw Status.INVALID_ARGUMENT
                    .withDescription("Unknown command type: ${cmd.typeUrl}")
                    .asRuntimeException()
            }

            val eventBookWithCover = eventBook.toBuilder()
                .setCover(cmdBook.cover)
                .build()

            return BusinessResponse.newBuilder()
                .setEvents(eventBookWithCover)
                .build()

        } catch (e: CommandValidationException) {
            throw Status.fromCode(e.statusCode)
                .withDescription(e.message)
                .asRuntimeException()
        }
    }
}
